server/util/rexec/rexec.go

```diff
@@ -0,0 +1,287 @@
+// package rexec provides utility functions for remote execution clients.
+package rexec
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/cachetools"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/digest"
+	"github.com/buildbuddy-io/buildbuddy/server/util/retry"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/genproto/googleapis/longrunning"
+
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	rspb "github.com/buildbuddy-io/buildbuddy/proto/resource"
+	gstatus "google.golang.org/grpc/status"
+)
+
+// MakeEnv assembles a list of EnvironmentVariable protos from a list of
+// NAME=VALUE pairs. If the same name is specified more than once, the last one
+// wins. The entries are sorted by name, so that the environment variables are
+// cache-friendly.
+func MakeEnv(pairs ...string) ([]*repb.Command_EnvironmentVariable, error) {
+	m, err := parsePairs(pairs)
+	if err != nil {
+		return nil, err
+	}
+	names := maps.Keys(m)
+	sort.Strings(names)
+	out := make([]*repb.Command_EnvironmentVariable, 0, len(m))
+	for _, name := range names {
+		out = append(out, &repb.Command_EnvironmentVariable{
+			Name:  name,
+			Value: m[name],
+		})
+	}
+	return out, nil
+}
+
+// MakePlatform assembles a Platform proto from a list of NAME=VALUE pairs. If
+// the same name is specified more than once, the last one wins. The entries are
+// sorted by name, so that the platform is cache-friendly.
+func MakePlatform(pairs ...string) (*repb.Platform, error) {
+	m, err := parsePairs(pairs)
+	if err != nil {
+		return nil, err
+	}
+	names := maps.Keys(m)
+	sort.Strings(names)
+	p := &repb.Platform{Properties: make([]*repb.Platform_Property, 0, len(names))}
+	for _, name := range names {
+		p.Properties = append(p.Properties, &repb.Platform_Property{
+			Name:  name,
+			Value: m[name],
+		})
+	}
+	return p, nil
+}
+
+// parsePairs parses a list of "NAME=VALUE" pairs into a map. If the same NAME
+// appears more than once, the last one wins.
+func parsePairs(pairs []string) (map[string]string, error) {
+	m := map[string]string{}
+	for _, pair := range pairs {
+		parts := strings.SplitN(pair, "=", 2)
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid environment variable %q (expected NAME=VALUE)", pair)
```

Comment:
would prefer to use a status error here rather than a fmt.Errorf

InvalidInput fits

Reply:
done